### ISSUE TITLE
tensorflow2.0生成的pb模型没法识别

### ISSUE THREAD

**Issue by xinyidaren** [alibaba/MNN#485]
Check failed: success ==> read_proto_from_binary failed

代码如下：
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import sklearn
import pandas as pd
import os
import sys
import time
import tensorflow as tf

from tensorflow import keras
from tensorflow.keras.callbacks import Callback

print(sys.version_info)
for module in mpl, np, pd, sklearn, tf, keras:
    print(module.__name__, module.__version__)
fashion_mnist = keras.datasets.fashion_mnist
(x_train_all, y_train_all), (x_test, y_test) = fashion_mnist.load_data()
x_valid, x_train = x_train_all[:5000], x_train_all[5000:]
y_valid, y_train = y_train_all[:5000], y_train_all[5000:]

print(x_valid.shape, y_valid.shape)
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

x_train = x_train.astype(np.float32).reshape(-1, 28, 28,  1)/255
input1 = keras.Input(shape=(28,28,1),dtype=tf.float32)
_x = keras.layers.Conv2D(64,(3,3),strides = 2,activation='relu',use_bias = False)(input1)
_x = keras.layers.Conv2D(32,(3,3),strides = 2,activation='relu',use_bias = False)(_x)
_x = keras.layers.Reshape((1,1,1152))(_x)
_x = keras.layers.Conv2D(256,(1,1),activation='relu',use_bias = False)(_x)
_x = keras.layers.Conv2D(10,(1,1),activation=None,use_bias = False)(_x)
_x = keras.layers.Reshape((10,))(_x)
output1 = keras.layers.Softmax()(_x)
model = keras.Model(inputs=[input1],outputs=[output1])
print(model)
model.summary()

model.compile(loss="sparse_categorical_crossentropy",
              optimizer = "sgd",
              metrics = ['accuracy'])
history = model.fit(x_train, y_train, epochs=3,batch_size=32)
tf.saved_model.save(model,"./keras_saved_graph")

**Comment by czy2014hust** [comment]
freeze the graph reference from the below code
```python
from keras import backend as K
import tensorflow as tf

def freeze_session(session, keep_var_names=None, output_names=None, clear_devices=True):
    """
    Freezes the state of a session into a pruned computation graph.

    Creates a new computation graph where variable nodes are replaced by
    constants taking their current value in the session. The new graph will be
    pruned so subgraphs that are not necessary to compute the requested
    outputs are removed.
    @param session The TensorFlow session to be frozen.
    @param keep_var_names A list of variable names that should not be frozen,
                          or None to freeze all the variables in the graph.
    @param output_names Names of the relevant graph outputs.
    @param clear_devices Remove the device directives from the graph for better portability.
    @return The frozen graph definition.
    """
    from tensorflow.python.framework.graph_util import convert_variables_to_constants
    graph = session.graph
    with graph.as_default():
        freeze_var_names = list(set(v.op.name for v in tf.global_variables()).difference(keep_var_names or []))
        output_names = output_names or []
        output_names += [v.op.name for v in tf.global_variables()]
        # Graph -> GraphDef ProtoBuf
        input_graph_def = graph.as_graph_def()
        if clear_devices:
            for node in input_graph_def.node:
                node.device = ""
        frozen_graph = convert_variables_to_constants(session, input_graph_def,
                                                      output_names, freeze_var_names)
        return frozen_graph


frozen_graph = freeze_session(K.get_session(),
                              output_names=[out.op.name for out in model.outputs]

tf.train.write_graph(frozen_graph, "model", "tf_model.pb", as_text=False)

**Comment by hahadashi** [comment]
> freeze the graph reference from the below code
> 
> ```python
> from keras import backend as K
> import tensorflow as tf
> 
> def freeze_session(session, keep_var_names=None, output_names=None, clear_devices=True):
>     """
>     Freezes the state of a session into a pruned computation graph.
> 
>     Creates a new computation graph where variable nodes are replaced by
>     constants taking their current value in the session. The new graph will be
>     pruned so subgraphs that are not necessary to compute the requested
>     outputs are removed.
>     @param session The TensorFlow session to be frozen.
>     @param keep_var_names A list of variable names that should not be frozen,
>                           or None to freeze all the variables in the graph.
>     @param output_names Names of the relevant graph outputs.
>     @param clear_devices Remove the device directives from the graph for better portability.
>     @return The frozen graph definition.
>     """
>     from tensorflow.python.framework.graph_util import convert_variables_to_constants
>     graph = session.graph
>     with graph.as_default():
>         freeze_var_names = list(set(v.op.name for v in tf.global_variables()).difference(keep_var_names or []))
>         output_names = output_names or []
>         output_names += [v.op.name for v in tf.global_variables()]
>         # Graph -> GraphDef ProtoBuf
>         input_graph_def = graph.as_graph_def()
>         if clear_devices:
>             for node in input_graph_def.node:
>                 node.device = ""
>         frozen_graph = convert_variables_to_constants(session, input_graph_def,
>                                                       output_names, freeze_var_names)
>         return frozen_graph
> 
> 
> frozen_graph = freeze_session(K.get_session(),
>                               output_names=[out.op.name for out in model.outputs]
> 
> tf.train.write_graph(frozen_graph, "model", "tf_model.pb", as_text=False)
> ```

hi，i try to convert tf2.0 ckpt model to the pb model，but the result between with them is different

**Comment by hahadashi** [comment]
> freeze the graph reference from the below code
> 
> ```python
> from keras import backend as K
> import tensorflow as tf
> 
> def freeze_session(session, keep_var_names=None, output_names=None, clear_devices=True):
>     """
>     Freezes the state of a session into a pruned computation graph.
> 
>     Creates a new computation graph where variable nodes are replaced by
>     constants taking their current value in the session. The new graph will be
>     pruned so subgraphs that are not necessary to compute the requested
>     outputs are removed.
>     @param session The TensorFlow session to be frozen.
>     @param keep_var_names A list of variable names that should not be frozen,
>                           or None to freeze all the variables in the graph.
>     @param output_names Names of the relevant graph outputs.
>     @param clear_devices Remove the device directives from the graph for better portability.
>     @return The frozen graph definition.
>     """
>     from tensorflow.python.framework.graph_util import convert_variables_to_constants
>     graph = session.graph
>     with graph.as_default():
>         freeze_var_names = list(set(v.op.name for v in tf.global_variables()).difference(keep_var_names or []))
>         output_names = output_names or []
>         output_names += [v.op.name for v in tf.global_variables()]
>         # Graph -> GraphDef ProtoBuf
>         input_graph_def = graph.as_graph_def()
>         if clear_devices:
>             for node in input_graph_def.node:
>                 node.device = ""
>         frozen_graph = convert_variables_to_constants(session, input_graph_def,
>                                                       output_names, freeze_var_names)
>         return frozen_graph
> 
> 
> frozen_graph = freeze_session(K.get_session(),
>                               output_names=[out.op.name for out in model.outputs]
> 
> tf.train.write_graph(frozen_graph, "model", "tf_model.pb", as_text=False)
> ```
Can I contact you by email?

**Comment by hahadashi** [comment]
> Check failed: success ==> read_proto_from_binary failed
> 
> 代码如下：
> import matplotlib as mpl
> import matplotlib.pyplot as plt
> import numpy as np
> import sklearn
> import pandas as pd
> import os
> import sys
> import time
> import tensorflow as tf
> 
> from tensorflow import keras
> from tensorflow.keras.callbacks import Callback
> 
> print(sys.version_info)
> for module in mpl, np, pd, sklearn, tf, keras:
> print(module.**name**, module.**version**)
> fashion_mnist = keras.datasets.fashion_mnist
> (x_train_all, y_train_all), (x_test, y_test) = fashion_mnist.load_data()
> x_valid, x_train = x_train_all[:5000], x_train_all[5000:]
> y_valid, y_train = y_train_all[:5000], y_train_all[5000:]
> 
> print(x_valid.shape, y_valid.shape)
> print(x_train.shape, y_train.shape)
> print(x_test.shape, y_test.shape)
> 
> x_train = x_train.astype(np.float32).reshape(-1, 28, 28, 1)/255
> input1 = keras.Input(shape=(28,28,1),dtype=tf.float32)
> _x = keras.layers.Conv2D(64,(3,3),strides = 2,activation='relu',use_bias = False)(input1)
> _x = keras.layers.Conv2D(32,(3,3),strides = 2,activation='relu',use_bias = False)(_x)
> _x = keras.layers.Reshape((1,1,1152))(_x)
> _x = keras.layers.Conv2D(256,(1,1),activation='relu',use_bias = False)(_x)
> _x = keras.layers.Conv2D(10,(1,1),activation=None,use_bias = False)(_x)
> _x = keras.layers.Reshape((10,))(_x)
> output1 = keras.layers.Softmax()(_x)
> model = keras.Model(inputs=[input1],outputs=[output1])
> print(model)
> model.summary()
> 
> model.compile(loss="sparse_categorical_crossentropy",
> optimizer = "sgd",
> metrics = ['accuracy'])
> history = model.fit(x_train, y_train, epochs=3,batch_size=32)
> tf.saved_model.save(model,"./keras_saved_graph")

is your pb model working？

**Comment by hahadashi** [comment]
> freeze the graph reference from the below code
> 
> ```python
> from keras import backend as K
> import tensorflow as tf
> 
> def freeze_session(session, keep_var_names=None, output_names=None, clear_devices=True):
>     """
>     Freezes the state of a session into a pruned computation graph.
> 
>     Creates a new computation graph where variable nodes are replaced by
>     constants taking their current value in the session. The new graph will be
>     pruned so subgraphs that are not necessary to compute the requested
>     outputs are removed.
>     @param session The TensorFlow session to be frozen.
>     @param keep_var_names A list of variable names that should not be frozen,
>                           or None to freeze all the variables in the graph.
>     @param output_names Names of the relevant graph outputs.
>     @param clear_devices Remove the device directives from the graph for better portability.
>     @return The frozen graph definition.
>     """
>     from tensorflow.python.framework.graph_util import convert_variables_to_constants
>     graph = session.graph
>     with graph.as_default():
>         freeze_var_names = list(set(v.op.name for v in tf.global_variables()).difference(keep_var_names or []))
>         output_names = output_names or []
>         output_names += [v.op.name for v in tf.global_variables()]
>         # Graph -> GraphDef ProtoBuf
>         input_graph_def = graph.as_graph_def()
>         if clear_devices:
>             for node in input_graph_def.node:
>                 node.device = ""
>         frozen_graph = convert_variables_to_constants(session, input_graph_def,
>                                                       output_names, freeze_var_names)
>         return frozen_graph
> 
> 
> frozen_graph = freeze_session(K.get_session(),
>                               output_names=[out.op.name for out in model.outputs]
> 
> tf.train.write_graph(frozen_graph, "model", "tf_model.pb", as_text=False)
> ```
@lldong @czy2014hust @nihui @llhe @xinyidaren 
hi，i use your demo complete frozen the ckpt model(checkpoint = tf.train.Checkpoint(myAwesomeModel=me), checkpoint.restore(tf.train.latest_checkpoint('examples'    ))) or h5 model (model.load_weights('model_enc.h5'))to pb model。after i run ckpt model and pb model，i find their result is different。and the result of ckpt model  is right。i check the pb model parmas， and find some op variable parmas is initial value，like “bias” is vector contain zero，“layer normal beta”  is vector contain zero eg. i think it is wrong,  hope your give me some advice,thx

**Comment by ShihuaiXu** [comment]
用TF1和TF2都不行
TF1显示有些op不能转
TF2初始化直接挂掉了
请问现在有解决方法了吗?
@czy2014hust
